apps/builder/app/routes/oauth.ws.authorize.tsx

```diff
@@ -0,0 +1,229 @@
+import { json, redirect, type LoaderFunction } from "@remix-run/server-runtime";
+import { z } from "zod";
+import createDebug from "debug";
+import { fromError } from "zod-validation-error";
+import { getAuthorizationServerOrigin } from "~/shared/router-utils/origins";
+import env from "~/env/env.server";
+import { authenticator } from "~/services/auth.server";
+import { returnToCookie } from "~/services/cookie.server";
+import { createCodeToken } from "~/services/token.server";
+import { isUserAuthorizedForProject } from "~/services/builder-access.server";
+
+const debug = createDebug("OAuth").extend("Authorize");
+
+// https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.2.1
+const createOauthError =
+  (redirectUri: string, state: string) =>
+  (
+    error: "invalid_request" | "invalid_scope" | "unauthorized_client",
+    errorDescription: string
+  ) => {
+    const url = new URL(redirectUri);
+    url.search = "";
+
+    url.searchParams.set("error", error);
+    url.searchParams.set("error_description", errorDescription);
+    if (state) {
+      url.searchParams.set("state", state);
+    }
+
+    return redirect(url.href, { status: 302 });
+  };
+
+/**
+ * OAuth 2.0 Authorization Request
+ *
+ * https://datatracker.ietf.org/doc/html/rfc7636#section-4.3
+ *
+ * https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.1
+ */
+const OAuthParamsSchema = z.object({
+  // Ensure that the response_type is valid and supported by the authorization server (e.g., code for the authorization code grant type).
+  response_type: z.literal("code"),
+  redirect_uri: z.string().url(),
+
+  client_id: z.string(),
+  state: z.string(),
+  scope: z
+    .string()
+    .refine((str) => str.startsWith("project:"), {
+      message: "Only 'project:' scopes are allowed",
+    })
+    .transform((scope) => ({
+      projectId: scope.split(":")[1],
+    })),
+  code_challenge: z.string(),
+  code_challenge_method: z.literal("S256"),
+});
+
+const OAuthRedirectUriSchema = z.object({
+  redirect_uri: z.string().url(),
+});
+
+/**
+ * Based RFC 6749 and RFC 7636
+ *
+ * https://datatracker.ietf.org/doc/html/rfc6749
+ *
+ * https://datatracker.ietf.org/doc/html/rfc7636
+ */
+export const loader: LoaderFunction = async ({ request }) => {
+  try {
```

Comment:
not in love with the giant try block, because its hard to see where it ends

Reply:
Most routers allows to intercept errors after the main logic without using try catch but not remix loader.